## DSCI401 - Homework 2

### Question 1

In [ ]:
# Importing the pandas library
import pandas as pd

In [ ]:
# Loading the dataset
teams = pd.read_csv("https://raw.githubusercontent.com/gjm112/DSCI401/main/data/Teams.csv")

In [ ]:
# Loading the first 5 rows
teams.head()

yearID lgID teamID franchID divID  Rank   G  Ghome   W   L  ...  DP     FP  \
0    1871  NaN    BS1      BNA   NaN     3  31    NaN  20  10  ...  24  0.834   
1    1871  NaN    CH1      CNA   NaN     2  28    NaN  19   9  ...  16  0.829   
2    1871  NaN    CL1      CFC   NaN     8  29    NaN  10  19  ...  15  0.818   
3    1871  NaN    FW1      KEK   NaN     7  19    NaN   7  12  ...   8  0.803   
4    1871  NaN    NY2      NNA   NaN     5  33    NaN  16  17  ...  14  0.840   

                      name                          park  attendance  BPF  \
0     Boston Red Stockings           South End Grounds I         NaN  103   
1  Chicago White Stockings       Union Base-Ball Grounds         NaN  104   
2   Cleveland Forest Citys  National Association Grounds         NaN   96   
3     Fort Wayne Kekiongas                Hamilton Field         NaN  101   
4         New York Mutuals      Union Grounds (Brooklyn)         NaN   90   

   PPF  teamIDBR  teamIDlahman45  teamIDretro  
0   98       BOS             BS1          BS1  
1  102       CHI             CH1          CH1  
2  100       CLE             CL1          CL1  
3  107       KEK             FW1          FW1  
4   88       NYU             NY2          NY2  

[5 rows x 48 columns]

In [ ]:
# Getting the shape of the
teams.shape

(2985, 48)

In [ ]:
# Getting the columns
teams.columns

Index(['yearID', 'lgID', 'teamID', 'franchID', 'divID', 'Rank', 'G', 'Ghome',
       'W', 'L', 'DivWin', 'WCWin', 'LgWin', 'WSWin', 'R', 'AB', 'H', 'X2B',
       'X3B', 'HR', 'BB', 'SO', 'SB', 'CS', 'HBP', 'SF', 'RA', 'ER', 'ERA',
       'CG', 'SHO', 'SV', 'IPouts', 'HA', 'HRA', 'BBA', 'SOA', 'E', 'DP', 'FP',
       'name', 'park', 'attendance', 'BPF', 'PPF', 'teamIDBR',
       'teamIDlahman45', 'teamIDretro'],
      dtype='object')

## (a)

In [ ]:
# Filtering and selecting specific rows and columns
teams_new = teams[(teams['yearID'] >= 2000) & (teams['yearID'] <= 2009)][['yearID', 'W', 'L']]
teams_new.head()

yearID   W   L
2325    2000  82  80
2326    2000  85  77
2327    2000  95  67
2328    2000  74  88
2329    2000  85  77

## (b)

In [ ]:
# filtering the teams data for HR>=200 for the Chicago cubs
cubs_HR = teams[(teams["teamID"] == "CHN") & (teams["HR"] >= 200)]

# finding the number of unique years the cubs had HR>=200
total_years = len(cubs_HR["yearID"].unique())
print("Number of years the Chicago Cubs hit at least 200 HRs:" ,total_years)

# finding the median of the wins
median_wins = cubs_HR["W"].median()
print("\nThe median number of wins in the seasons the Chicago Cubs hit at least 200 HRs:" ,median_wins)

Number of years the Chicago Cubs hit at least 200 HRs: 7

The median number of wins in the seasons the Chicago Cubs hit at least 200 HRs: 84.0


## (c)

In [ ]:
# creating the factor election based on the presidential terms
teams["election"] = ((teams["yearID"]-1788) // 4) + 1
teams.head()

yearID lgID teamID franchID divID  Rank   G  Ghome   W   L  ...     FP  \
0    1871  NaN    BS1      BNA   NaN     3  31    NaN  20  10  ...  0.834   
1    1871  NaN    CH1      CNA   NaN     2  28    NaN  19   9  ...  0.829   
2    1871  NaN    CL1      CFC   NaN     8  29    NaN  10  19  ...  0.818   
3    1871  NaN    FW1      KEK   NaN     7  19    NaN   7  12  ...  0.803   
4    1871  NaN    NY2      NNA   NaN     5  33    NaN  16  17  ...  0.840   

                      name                          park attendance  BPF  PPF  \
0     Boston Red Stockings           South End Grounds I        NaN  103   98   
1  Chicago White Stockings       Union Base-Ball Grounds        NaN  104  102   
2   Cleveland Forest Citys  National Association Grounds        NaN   96  100   
3     Fort Wayne Kekiongas                Hamilton Field        NaN  101  107   
4         New York Mutuals      Union Grounds (Brooklyn)        NaN   90   88   

   teamIDBR  teamIDlahman45  teamIDretro  election  
0       BOS             BS1          BS1        21  
1       CHI             CH1          CH1        21  
2       CLE             CL1          CL1        21  
3       KEK             FW1          FW1        21  
4       NYU             NY2          NY2        21  

[5 rows x 49 columns]

In [ ]:
# finding the home runs for each 4 year term
HR_term = teams.groupby('election')['HR'].sum().reset_index()
HR_term.head()

election    HR
0        21    47
1        22   164
2        23   145
3        24   554
4        25  2031

In [ ]:
# Finding the presidential term with the most home runs
HR_term_max = HR_term[HR_term['HR'] == HR_term['HR'].max()]
HR_term_max

election     HR
37        58  24076

### The presidential term with the most home runs is the 58th preseidential term with a total of 24076 home runs

## (d)

In [ ]:
# loading the plotly package
import plotly.express as px

# removing observations where league ID is missing
teams_lgID_complete = teams.dropna(subset=['lgID'])

# Creating a line plot using Plotly Express
fig = px.line(teams_lgID_complete, x = 'yearID', y = 'HR', color = 'lgID',
              labels = {'yearID': 'Year', 'HR': 'Total Home Runs'},
              title ='Total Home Runs Per Season Stratified by League')
fig.show()

## (e)

In [ ]:
# creating the indicator variable called winning record
teams['winning_record'] = teams['W'] > teams['L']
teams.head()

yearID lgID teamID franchID divID  Rank   G  Ghome   W   L  ...  \
0    1871  NaN    BS1      BNA   NaN     3  31    NaN  20  10  ...   
1    1871  NaN    CH1      CNA   NaN     2  28    NaN  19   9  ...   
2    1871  NaN    CL1      CFC   NaN     8  29    NaN  10  19  ...   
3    1871  NaN    FW1      KEK   NaN     7  19    NaN   7  12  ...   
4    1871  NaN    NY2      NNA   NaN     5  33    NaN  16  17  ...   

                      name                          park attendance  BPF  PPF  \
0     Boston Red Stockings           South End Grounds I        NaN  103   98   
1  Chicago White Stockings       Union Base-Ball Grounds        NaN  104  102   
2   Cleveland Forest Citys  National Association Grounds        NaN   96  100   
3     Fort Wayne Kekiongas                Hamilton Field        NaN  101  107   
4         New York Mutuals      Union Grounds (Brooklyn)        NaN   90   88   

   teamIDBR  teamIDlahman45  teamIDretro  election  winning_record  
0       BOS             BS1          BS1        21            True  
1       CHI             CH1          CH1        21            True  
2       CLE             CL1          CL1        21           False  
3       KEK             FW1          FW1        21           False  
4       NYU             NY2          NY2        21           False  

[5 rows x 50 columns]

In [ ]:
# Creating a line plot using Plotly Express
fig = px.scatter(teams, x = 'R', y = 'RA', color = 'winning_record',
              labels = {'R': 'Runs (R)', 'RA': 'Runs Against (RA)'},
              title ='Scatter Plot of Runs (R) vs Runs against (RA)')
fig.show()

# Question 2

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [ ]:
# Load the "Violations" dataset
violations = pd.read_csv('/content/drive/MyDrive/MS DATA SCIENCE/Violations.csv')

In [ ]:
# loading the first 5 rows
violations.head()

camis                    dba   boro  building           street  zipcode  \
0  30075445  MORRIS PARK BAKE SHOP  BRONX    1007.0  MORRIS PARK AVE    10462   
1  30075445  MORRIS PARK BAKE SHOP  BRONX    1007.0  MORRIS PARK AVE    10462   
2  30075445  MORRIS PARK BAKE SHOP  BRONX    1007.0  MORRIS PARK AVE    10462   
3  30075445  MORRIS PARK BAKE SHOP  BRONX    1007.0  MORRIS PARK AVE    10462   
4  30075445  MORRIS PARK BAKE SHOP  BRONX    1007.0  MORRIS PARK AVE    10462   

          phone inspection_date  \
0  7.188925e+09      2015-02-09   
1  7.188925e+09      2014-03-03   
2  7.188925e+09      2013-10-10   
3  7.188925e+09      2013-09-11   
4  7.188925e+09      2013-09-11   

                                              action violation_code  score  \
0    Violations were cited in the following area(s).            06C    6.0   
1    Violations were cited in the following area(s).            10F    2.0   
2  No violations were recorded at the time of thi...            NaN    NaN   
3    Violations were cited in the following area(s).            04L    6.0   
4    Violations were cited in the following area(s).            04N    6.0   

  grade  grade_date record_date                           inspection_type  \
0     A  2015-02-09  2016-01-06     Cycle Inspection / Initial Inspection   
1     A  2014-03-03  2016-01-06     Cycle Inspection / Initial Inspection   
2   NaN         NaN  2016-01-06  Trans Fat / Second Compliance Inspection   
3     A  2013-09-11  2016-01-06          Cycle Inspection / Re-inspection   
4     A  2013-09-11  2016-01-06          Cycle Inspection / Re-inspection   

   cuisine_code  
0             8  
1             8  
2             8  
3             8  
4             8

In [ ]:
# getting the shape
violations.shape

(480621, 16)

In [ ]:
# getting the columns
violations.columns

Index(['camis', 'dba', 'boro', 'building', 'street', 'zipcode', 'phone',
       'inspection_date', 'action', 'violation_code', 'score', 'grade',
       'grade_date', 'record_date', 'inspection_type', 'cuisine_code'],
      dtype='object')

## (a)

In [ ]:
# Filling the missing values in the grade column as Missing
violations['grade'] = violations['grade'].fillna('Missing')

# Calculating the count of inspections with a grade of 'A' and grouping by boro.
inspection_A = violations[violations['grade'] == 'A'].groupby('boro')['grade'].count()

# Calculating the total count of inspections of grade and grouping by boror
total_inspections = violations.groupby('boro')['grade'].count()

# Calculate the proportion
proportion_A = inspection_A / total_inspections

# Creating a DataFrame to display the column name Proportion_A
result_df = pd.DataFrame({'Proportion_A': proportion_A})

result_df

Proportion_A
boro                       
BRONX              0.351805
BROOKLYN           0.329875
MANHATTAN          0.342472
Missing            0.452055
QUEENS             0.328954
STATEN ISLAND      0.362419

## (b)

In [ ]:
# Finding the top ten dba's with the most number of inspections
top_dba = violations['dba'].value_counts().nlargest(10).index.tolist()

# Filtering to keep only the top ten DBAs
top10_dba = violations[violations['dba'].isin(top_dba)]

# Computing the average score for each DBA
dba_mean_score = top10_dba.groupby('dba')['score'].mean().reset_index()

# Sorting by mean score in ascending order
mean_score_sorted = dba_mean_score.sort_values(by='score')

mean_score_sorted

dba      score
8                              STARBUCKS  11.664354
3                         DUNKIN' DONUTS  13.597921
4         DUNKIN' DONUTS, BASKIN ROBBINS  14.450216
9                                 SUBWAY  14.806099
7                             MCDONALD'S  16.977680
0                            BURGER KING  17.337428
6                  KENNEDY FRIED CHICKEN  18.042160
5  GOLDEN KRUST CARIBBEAN BAKERY & GRILL  18.276753
1                    CROWN FRIED CHICKEN  18.562773
2                               DOMINO'S  18.635636

### STARBUCKS had the lowest average inspection score

## (c)

In [ ]:
# Filter for zip codes in Manhattan with 50 or more inspections
manhattan_zips = violations[(violations['boro'] == 'MANHATTAN')]
manhattan_zips = manhattan_zips.groupby('zipcode').filter(lambda x: len(x) >= 50)

# Calculate the median violation score by zip code
median_scores = manhattan_zips.groupby('zipcode')['score'].median().reset_index()

# Print the median scores
median_scores.head()

zipcode  score
0    10001   15.0
1    10002   18.0
2    10003   17.0
3    10004   14.0
4    10005   17.0

### From observations, there appears to be an inverse relationship between the median score and number of inspections. The zip codes that had more inspections had lower median score. This could be due that the zip codes with more inspections complied with regulations.